### DISCUSSION TITLE
[FEATURE REQUEST] Configurable filename format (with code provided)

### DISCUSSION THREAD

**Discussion by Michoko92** [hlky/stable-diffusion-webui#504] [Feature suggestions]
Hi,

All parameters vary depending on our needs and experiments. Some people play a lot with steps, others with CFG, or image dimensions. And it is very handy to have this information in the filenames. Personally, I have to manually change webui.py every time I update the repo, to fit my needs. The best way to make everyone happy would be to make the filenames format configurable with a "--name_format" argument. Using tags should be the most flexible and scalable way. Something like:

`--name_format [STEPS]-[CFG]-[SEED]-[PROMPT]`

Here is the code that implements this (sorry I'm not familiar with Github, and I don't want to mess up things):

First add the following argument in the arguments list:

`parser.add_argument("--name_format", type=str, nargs="?", help="filenames format", default=None)`

Then replace this code in webui.py :

```
for i, x_sample in enumerate(x_samples_ddim):
                sanitized_prompt = prompts[i].replace(' ', '_').translate({ord(x): '' for x in invalid_filename_chars})
                if sort_samples:
                    sanitized_prompt = sanitized_prompt[:128] #200 is too long
                    sample_path_i = os.path.join(sample_path, sanitized_prompt)
                    os.makedirs(sample_path_i, exist_ok=True)
                    base_count = get_next_sequence_number(sample_path_i)
                    filename = f"{base_count:05}-{steps}_{sampler_name}_{seeds[i]}"
                else:
                    sample_path_i = sample_path
                    base_count = get_next_sequence_number(sample_path_i)
                    sanitized_prompt = sanitized_prompt
                    filename = f"{base_count:05}-{steps}_{sampler_name}_{seeds[i]}_{sanitized_prompt}"[:128] #same as before
```

With this code:

```
            for i, x_sample in enumerate(x_samples_ddim):
                sanitized_prompt = prompts[i].replace(' ', '_').translate({ord(x): '' for x in invalid_filename_chars})
                if sort_samples:
                    sanitized_prompt = sanitized_prompt[:128] #200 is too long
                    sample_path_i = os.path.join(sample_path, sanitized_prompt)
                    os.makedirs(sample_path_i, exist_ok=True)
                    base_count = get_next_sequence_number(sample_path_i)
                    filename = opt.name_format or "[STEPS]-[SAMPLER]-[SEED]"
                else:
                    sample_path_i = sample_path
                    base_count = get_next_sequence_number(sample_path_i)
                    sanitized_prompt = sanitized_prompt
                    filename = opt.name_format or "[STEPS]-[SAMPLER]-[SEED]-[PROMPT]"

                filename = f"{base_count:05}-" + filename
                filename = filename.replace("[STEPS]", str(steps))
                filename = filename.replace("[CFG]", str(cfg_scale))
                filename = filename.replace("[PROMPT]", sanitized_prompt[:128])
                filename = filename.replace("[WIDTH]", str(width))
                filename = filename.replace("[HEIGHT]", str(height))
                filename = filename.replace("[SAMPLER]", sampler_name)
                filename = filename.replace("[SEED]", str(seeds[i]))
```

I made some tests, and it seems to work fine. As you can see, the supported tags in the name format are: [STEPS], [CFG], [PROMPT], [WIDTH], [HEIGHT], [SAMPLER] and [SEED]. If no format is provided as an argument, it will use the current naming format.

Hope it will help. Keep up the great work!

**Comment by Michoko92** [comment]
I updated the code above, as I realized the "count" variable had to always be at the beginning of the filename to work properly, and be followed by a "-". So it's better not to make it configurable, and I removed the "[COUNT]" tag.

**Comment by wmooring** [comment]
Agreed with this, also I think it would be a good idea if the cfg value were included in the filename by default (I added "c{cfg_scale}" to mine after the seed value, to differentiate it from the step count), which would make it so the generation is reproducable with nothing but the image file.  Also, a toggle of whether or not to replace the spaces in the prompt with underscores in the filenames would be nice (I modified the "sanitized_prompt" code to disable that in mine, since I want to be able to copy and paste the original prompt more easily from the filename, and underscores *do* affect the prompt differently than spaces when generating)

**Comment by Michoko92** [comment]
Finally made a PR here: https://github.com/sd-webui/stable-diffusion-webui/pull/908

**Comment by thedyze** [comment]
Agree. I also manually update to include guidance and denoise values..
